### PR TITLE
HW5. Redux

### PR DESCRIPTION
**Pull request by vm137**

### CODE REVIEW

File: cypress/plugins/index.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // ***********************************************************
-// This example plugins/index.js can be used to load plugins
+// This example plugins/rootReducer.js can be used to load plugins
```
Comment (loginov-rocks) [inline cypress/plugins/index.js]: Unintended change by auto refactor.

File: cypress/support/index.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 // ***********************************************************
-// This example support/index.js is processed and
+// This example support/rootReducer.js is processed and
```
Comment (loginov-rocks) [inline cypress/support/index.js]: Unintended change by auto refactor.

File: package.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
     "lint": "eslint --ext js,jsx src",
     "test": "jest src",
     "test:cover": "jest src --coverage",
+    "test:watch": "jest src -- --watch",
```
Comment (loginov-rocks) [inline package.json]: Unnecessary `--`, there can be one of the two options:
* `npm test -- --watch`
* `jest src --watch`

File: package.json [hunk 4]
```diff
@@ -57,7 +61,9 @@
     "html-webpack-plugin": "^3.2.0",
     "identity-obj-proxy": "^3.0.0",
     "jest": "^24.5.0",
+    "moxios": "^0.4.0",
     "node-sass": "^4.11.0",
+    "redux-devtools-extension": "^2.13.8",
```
Comment (loginov-rocks) [inline package.json]: Unused dependency.

File: src/client/constants/constants.js [hunk 5]
```diff
@@ -1,6 +1,17 @@
-const COLOR = {
+/* action types */
```
Comment (loginov-rocks) [inline src/client/constants/constants.js]: `ACTIONS` are actions **types** in essence, it's a common approach to locate them next to the actions/action creators/thunks. So it'd be better to move they to `src/client/actions/types.js` for example.

File: src/client/actions/actions.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import { ACTIONS } from '../constants/constants';
+
+/* action creators */
+
+/* Movies */
+export const showSearchBlockAction = () => ({
+  type: ACTIONS.SHOW_SEARCH_BLOCK,
+});
+
+export const PropagateMoviesAction = data => ({
```
Comment (loginov-rocks) [inline src/client/actions/actions.js:11]: Why Uppercase?

File: src/client/actions/actions.test.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      moxios.stubRequest('http://react-cdp-api.herokuapp.com/movies?search=searchString&searchBy=&offset=&limit=', {
+        status: 200,
+        responseText: 'data',
+      });
+
+      const respFn = Action.fetchMoviesAction('searchString');
+      respFn(dispatch);
+
+      moxios.wait(() => {
+        expect(dispatch.mock.calls.length).toEqual(1);
```
Comment (loginov-rocks) [inline src/client/actions/actions.test.js]: Check what has been dispatched like so:
```js
expect(dispatch.mock.calls[0]).toEqual([{ payload: 'data', type: 'PROPAGATE_MOVIES' }]);
```

File: src/client/actions/actions.test.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        status: 200,
+        responseText: 'data',
+      });
+
+      const movieId = 1;
+      const respFn = Action.fetchSingleMovieAction(movieId);
+      respFn(dispatch);
+
+      moxios.wait(() => {
+        expect(dispatch.mock.calls.length).toEqual(2);
```
Comment (loginov-rocks) [inline src/client/actions/actions.test.js]: Check what has been dispatched like in the example above.

File: src/client/components/MainWindow/index.jsx [hunk 9]
```diff
@@ -1,3 +1,8 @@
+import { connect } from 'react-redux';
 import MainWindow from './MainWindow';
 
-export default MainWindow;
+const mapStateToProps = state => ({
```
Comment (loginov-rocks) [inline src/client/components/MainWindow/index.jsx]: You can also (not required) use short record with destructuring:
```js
const mapStateToProps = ({ showPage }) => ({ showPage });
```

File: src/client/components/MovieCard/index.jsx [hunk 10]
```diff
@@ -1,3 +1,9 @@
+import { connect } from 'react-redux';
 import MovieCard from './MovieCard';
+import { fetchSingleMovieAction } from '../../actions/actions';
 
-export default MovieCard;
+const mapDispatchProps = dispatch => ({
```
Comment (loginov-rocks) [inline src/client/components/MovieCard/index.jsx]: You can also (not required) use short record:
```js
const mapDispatchToProps = { fetchSingleMovie: fetchSingleMovieAction };
```

File: src/client/components/MovieCard/index.test.jsx [hunk 11]
```diff
@@ -0,0 +1,6 @@
+
+describe('MovieCard index.jsx', () => {
```
Comment (loginov-rocks) [inline src/client/components/MovieCard/index.test.jsx]: Please clean up.

File: src/client/components/SingleMovie/SingleMovie.test.jsx [hunk 12]
```diff
@@ -1,38 +1,40 @@
...
+      showSearchBlock={onClickJest}
     />);
 
     const button = wrapper.find('.btn-search');
     expect(button.length).toBe(1);
     button.simulate('click');
-    expect(onClickJest.mock.calls.length).toEqual(1);
 
-    expect(global.fetch).toHaveBeenCalledWith('http://react-cdp-api.herokuapp.com/movies/336');
+    // expect(onClickJest.mock.calls.length).toEqual(1);
```
Comment (loginov-rocks) [inline src/client/components/SingleMovie/SingleMovie.test.jsx]: Please clean up unused code and make skipped tests working. Here we need to check that `showSearchBlock` was triggered on `click` simulation.

**Review by loginov-rocks — CHANGES_REQUESTED** [review]

File: cypress/plugins/index.js (hunk 1 above)
Comment (vm137) [reply cypress/plugins/index.js]: fixed

File: cypress/support/index.js (hunk 2 above)
Comment (vm137) [reply cypress/support/index.js]: fixed.

File: package.json (hunk 3 above)
Comment (vm137) [reply package.json]: fixed

File: package.json (hunk 4 above)
Comment (vm137) [reply package.json]: removed

File: src/client/constants/constants.js (hunk 5 above)
Comment (vm137) [reply src/client/constants/constants.js]: fixed

File: src/client/actions/actions.js (hunk 6 above)
Comment (vm137) [reply src/client/actions/actions.js:11]: fixed to "actions"

File: src/client/actions/actions.test.js (hunk 7 above)
Comment (vm137) [reply src/client/actions/actions.test.js]: wow! this is real progress with async testing! 
done!

File: src/client/actions/actions.test.js (hunk 8 above)
Comment (vm137) [reply src/client/actions/actions.test.js]: done.

File: src/client/components/MainWindow/index.jsx (hunk 9 above)
Comment (vm137) [reply src/client/components/MainWindow/index.jsx]: I don't understand how we use it here. but it works. 
done according to your comment.

File: src/client/components/MovieCard/index.jsx (hunk 10 above)
Comment (vm137) [reply src/client/components/MovieCard/index.jsx]: I don't understand how we can omit "dispatch" here??
done as you proposed. ty

File: src/client/components/MovieCard/index.test.jsx (hunk 11 above)
Comment (vm137) [reply src/client/components/MovieCard/index.test.jsx]: do you mean first empty line?
if yes, then done.

File: src/client/components/SingleMovie/SingleMovie.test.jsx (hunk 12 above)
Comment (vm137) [reply src/client/components/SingleMovie/SingleMovie.test.jsx]: done.

File: src/client/components/MovieCard/index.test.jsx (hunk 11 above)
Comment (loginov-rocks) [reply src/client/components/MovieCard/index.test.jsx]: This file is useless :)

Comment (vm137) [reply src/client/components/MovieCard/index.test.jsx]: yes, I agree.

**Review by loginov-rocks — APPROVED** [review]
LGTM

**Pull request merged**